cmake/ThrustBuildCompilerTargets.cmake

```diff
@@ -20,27 +31,45 @@ function(thrust_build_compiler_targets)
   if (THRUST_TBB_FOUND)
     # There's a ton of these in the TBB backend, even though the code is correct.
     # TODO: silence these warnings in code instead
-    append_option_if_available("-Wno-unused-parameter" cxx_compile_options)
+#    append_option_if_available("-Wno-unused-parameter" cxx_compile_options)
   endif()
 
   if ("MSVC" STREQUAL "${CMAKE_CXX_COMPILER_ID}")
-    # TODO Enable /Wall instead of W3
-    append_option_if_available("/W3" cxx_compile_options)
+    append_option_if_available("/W4" cxx_compile_options)
+
+    # Treat all warnings as errors. This is only supported on Release builds,
+    # as `nv_exec_check_disable` doesn't seem to work with MSVC debug iterators
+    # and spurious warnings are emitted.
+    # See NVIDIA/thrust#1273, NVBug 3129879.
+    if (CMAKE_BUILD_TYPE STREQUAL "Release")
+      append_option_if_available("/WX" cxx_compile_options)
+    endif()
 
-    # Treat all warnings as errors:
-    append_option_if_available("/WX" cxx_compile_options)
+    # Suppress overly-pedantic/unavoidable warnings brought in with /W4:
+    # C4324: structure was padded due to alignment specifier
+    append_option_if_available("/wd4324" cxx_compile_options)
+    # C4505: unreferenced local function has been removed
+    # The CUDA `host_runtime.h` header emits this for
+    # `__cudaUnregisterBinaryUtil`.
+    append_option_if_available("/wd4505" cxx_compile_options)
+    # C4706: assignment within conditional expression
+    # MSVC doesn't provide an opt-out for this warning when the assignment is
+    # intentional. Clang will warn for these, but suppresses the warning when
+    # double-parentheses are used around the assignment. We'll let Clang catch
+    # unintentional assignments and suppress all such warnings on MSVC.
+    append_option_if_available("/wd4706" cxx_compile_options)
 
     # Disabled loss-of-data conversion warnings.
     # TODO Re-enable.
-    append_option_if_available("/wd4244" cxx_compile_options)
-    append_option_if_available("/wd4267" cxx_compile_options)
+#    append_option_if_available("/wd4244" cxx_compile_options)
+#    append_option_if_available("/wd4267" cxx_compile_options)
```

Comment:
Any reason to leave these in the code instead of removing them?

Reply:
I'm removing the commented out flags once I have a clean build without them. I'm leaving them commented out in the draft so I can add them back if needed.